### PR TITLE
PUB-358 Poirazi & Mel Neuron Model

### PR DESCRIPTION
**Pull request by MaxASchwarzer**
Add code implementing the neuron model from Poirazi & Mel's paper _Impact of Active Dendrites and Structural Plasticity on the Memory Capacity of Neural Tissue_, and a number of accompanying experiments, for use in the SDR paper.

### CODE REVIEW

File: projects/poirazi_neuron_model/data_tools.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
+import numpy
```
Comment (subutai) [inline projects/poirazi_neuron_model/data_tools.py]: Can you put our copyright at the top of each file?

File: projects/poirazi_neuron_model/run_noise_experiment.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    Results are saved to the file pm_noise_FN.txt.
+    """
+    for noise in test_noise_levels:
+
+        fps = []
+        fns = []
+
+        for trial in range(num_trials):
+
+            successful_initialization = False
```
Comment (subutai) [inline projects/poirazi_neuron_model/run_noise_experiment.py]: Could just initialize error to 0 and do "while error == 0:"

File: projects/poirazi_neuron_model/run_tm_dim_experiment.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                          training_iters = 1,
+                          automatic_threshold = True,
+                          save_results = True):
+  """
+  Run an experiment tracking the performance of the temporal memory given
+  different input dimensions.  The number of active cells is kept fixed, so we
+  are in effect varying the sparsity of the input.   We track performance by
+  comparing the cells predicted to be active with the cells actually active in
+  the sequence without noise at every timestep, and averaging across timesteps.
+  Three metrics are used, correlation (Pearson's r, by numpy.corrcoef),
```
Comment (subutai) [inline projects/poirazi_neuron_model/run_tm_dim_experiment.py]: Shouldn't we just count the number of overlapping bits?

File: htmresearch/frameworks/poirazi_neuron_model/data_tools.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy
+from nupic.bindings.math import *
+
+def apply_noise(data, noise):
+    """
+    Applies noise to a sparse matrix.  Noise should be an integer between 0 and
```
Comment (subutai) [inline htmresearch/frameworks/poirazi_neuron_model/data_tools.py]: Should we just make noise a float between 0 and 1?

File: htmresearch/frameworks/poirazi_neuron_model/data_tools.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    100, indicating the percentage of ones in the original input to move.
+    The input matrix is modified in-place, and nothing is returned.
+    This operation does not affect the sparsity of the matrix, or of any
+    individual datapoint.
+    """
+
+    for i in range(data.nRows()):
+        ones = data.rowNonZeros(i)[0]
+        replace_indices = numpy.random.choice(ones, size=int(1.0*len(ones)*noise/100), replace = False)
```
Comment (subutai) [inline htmresearch/frameworks/poirazi_neuron_model/data_tools.py]: `/100.0`

File: htmresearch/frameworks/poirazi_neuron_model/data_tools.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	"""
+	Generates a set of data drawn from a uniform distribution.  The binning structure from Poirazi & Mel is
+	ignored, and all (dim choose num_active) arrangements are possible.  num_negatives samples are put into
+	a separate negatives category for output compatibility with generate_data, but are otherwise identical.
+
+	"""
+
+	sparse_data = [numpy.random.choice(dim, size = num_active, replace = False) for i in range(num_samples)]
+	data = [[0 for i in range(dim)] for i in range(num_samples)]
+	for datapoint, sparse_datapoint in zip(data, sparse_data):
```
Comment (subutai) [inline htmresearch/frameworks/poirazi_neuron_model/data_tools.py]: Should use indentation of two spaces throughout, per our code conventions.

File: htmresearch/frameworks/poirazi_neuron_model/neuron_model.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+        for row in range(self.num_dendrites):
+            synapses = numpy.random.choice(self.dim, self.dendrite_length, replace = False)
+            for synapse in synapses:
+                self.dendrites[synapse, row] = 1
+
+
+    def initialize_permanences(self):
+        self.permanences = copy.deepcopy(self.dendrites)
+        self.permanences = self.permanences*0.3
```
Comment (subutai) [inline htmresearch/frameworks/poirazi_neuron_model/neuron_model.py]: Shouldn't this 0.3 be self.initial_permanence?

File: htmresearch/frameworks/poirazi_neuron_model/neuron_model.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    def calculate_activation(self, datapoint):
+        """
+        Only for a single datapoint
+        """
+
+        activations = datapoint * self.dendrites
+        activations = self.nonlinearity(activations)
+        return activations.sum()
+
+    def choose_substitute_pool(self):
```
Comment (subutai) [inline htmresearch/frameworks/poirazi_neuron_model/neuron_model.py]: Is this used anywhere?

File: htmresearch/frameworks/poirazi_neuron_model/neuron_model.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    def __init__(self,
+                 size = 10000,
+                 num_dendrites = 1000,
+                 dendrite_length = 10,
+                 dim = 400,
+                 nonlinearity = threshold_nonlinearity(6),
+                 num_replacement = 25,
+                 initial_permanence = 0.5,
+                 permanence_threshold = 0.15,
+                 perm_dec = 0.0125,
```
Comment (subutai) [inline htmresearch/frameworks/poirazi_neuron_model/neuron_model.py]: Use full names, like `permanence_decrement`

File: htmresearch/frameworks/poirazi_neuron_model/neuron_model.py (hunk 8 above)
Comment (MaxASchwarzer) [inline htmresearch/frameworks/poirazi_neuron_model/neuron_model.py]: Oops, it's not.  That's a remnant of the old P&M learning rule implementation, which I thought I had completely scrubbed from this code.

File: projects/poirazi_neuron_model/run_tm_dim_experiment.py (hunk 3 above)
Comment (MaxASchwarzer) [inline projects/poirazi_neuron_model/run_tm_dim_experiment.py]: I also tested that metric (it is actually still gathered in the experiments, under the name similarity, using a simple Jaccard index calculation), but the issue is that it does not necessarily hold up well in conditions of low sparsity.  The number of overlapping bits can be quite high in those cases without there actually being any correlation whatsoever between what the system is predicting and what the true answer is.

File: projects/poirazi_neuron_model/run_noise_experiment.py (hunk 2 above)
Comment (MaxASchwarzer) [inline projects/poirazi_neuron_model/run_noise_experiment.py]: I'm not sure the logic here would actually work out, given that I only want to re-initialize if there's a problem.

**Comment by MaxASchwarzer** [comment]
@subutai Please review -- I think I've addressed all your comments.

File: projects/poirazi_neuron_model/run_tm_dim_experiment.py (hunk 3 above)
Comment (MaxASchwarzer) [inline projects/poirazi_neuron_model/run_tm_dim_experiment.py]: Note for posterity: We have now switched to using this metric, after in-person discussions.

File: projects/poirazi_neuron_model/run_tm_dim_experiment.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
+# ----------------------------------------------------------------------
```
Comment (subutai) [inline projects/poirazi_neuron_model/run_tm_dim_experiment.py]: Why are these TM files in `poirazi_neuron_model`?  Do you want to just rename the directory to `sdr_paper_neuron_models` or something like that?

**Comment by MaxASchwarzer** [comment]
Per our conversation, I've moved everything from this project into two folders in the sdr_paper project for the merge.

**Pull request merged**